### PR TITLE
Add CFSet

### PR DESCRIPTION
**Pull request by burtonageo**

### CODE REVIEW

File: src/set.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+pub type CFSetRef = *const __CFSet;
+pub type CFMutableSetRef = *mut __CFSet;
+
+pub type CFSetApplierFunction = extern "C" fn(value: *const c_void, context: *mut c_void);
+pub type CFSetCopyDescriptionCallBack = extern "C" fn(value: *const c_void) -> CFStringRef;
+pub type CFSetEqualCallBack = extern "C" fn(value1: *const c_void, value2: *const c_void) -> Boolean;
+pub type CFSetHashCallBack = extern "C" fn(value: *const c_void) -> CFHashCode;
+pub type CFSetReleaseCallBack = extern "C" fn(allocator: CFAllocatorRef, value: *const c_void);
+pub type CFSetRetainCallBack = extern "C" fn(allocator: CFAllocatorRef, value: *const c_void);
```
Comment (dcuddeback) [inline src/set.rs]: This function returns `const void *`:

``` c
typedef const void *    (*CFSetRetainCallBack)(CFAllocatorRef allocator, const void *value);
```

File: src/set.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+extern {
+    pub fn CFSetCreate(allocator: CFAllocatorRef, values: *const *mut c_void, numValues: CFIndex,
+                       callBacks: *const CFSetCallBacks) -> CFSetRef;
+    pub fn CFSetCreateCopy(allocator: CFAllocatorRef, theSet: CFSetRef) -> CFSetRef;
+
+    pub fn CFSetCreateMutable(allocator: CFAllocatorRef, capacity: CFIndex, callBacks: *const CFSetCallBacks)
+                              -> CFMutableSetRef;
+    pub fn CFSetCreateMutableCopy(allocator: CFAllocatorRef, capacity: CFIndex, theSet: CFSetRef) -> CFSetRef;
```
Comment (dcuddeback) [inline src/set.rs]: This function returns `CFMutableSetRef`:

``` c
CFMutableSetRef CFSetCreateMutableCopy(CFAllocatorRef allocator, CFIndex capacity, CFSetRef theSet);
```

File: src/set.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    pub fn CFSetContainsValue(theSet: CFSetRef, value: *const c_void) -> Boolean;
+    pub fn CFSetGetCount(theSet: CFSetRef) -> CFIndex;
+    pub fn CFSetGetCountOfValue(theSet: CFSetRef, value: *const c_void) -> CFIndex;
+    pub fn CFSetGetValue(theSet: CFSetRef, value: *const c_void) -> *const c_void;
+    pub fn CFSetGetValueIfPresent(theSet: CFSetRef, candidate: *const c_void, value: *const *mut c_void) -> Boolean;
```
Comment (dcuddeback) [inline src/set.rs]: `const` and `mut` should be reversed here. `value` is an out-parameter that will be written to with a value of a `*const c_void`.

File: src/set.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    pub version: CFIndex,
+    pub retain: CFSetRetainCallBack,
+    pub release: CFSetReleaseCallBack,
+    pub copyDescription: CFSetCopyDescriptionCallBack,
+    pub equal: CFSetEqualCallBack,
+    pub hash: CFSetHashCallBack
+}
+
+extern {
+    pub fn CFSetCreate(allocator: CFAllocatorRef, values: *const *mut c_void, numValues: CFIndex,
```
Comment (dcuddeback) [inline src/set.rs]: `const` and `mut` should be reversed here, too. The parameter is defined as `const void **values`, which is a mutable pointer to a const pointer.

File: src/set.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    pub fn CFSetCreateMutable(allocator: CFAllocatorRef, capacity: CFIndex, callBacks: *const CFSetCallBacks)
+                              -> CFMutableSetRef;
+    pub fn CFSetCreateMutableCopy(allocator: CFAllocatorRef, capacity: CFIndex, theSet: CFSetRef) -> CFSetRef;
+
+    pub fn CFSetContainsValue(theSet: CFSetRef, value: *const c_void) -> Boolean;
+    pub fn CFSetGetCount(theSet: CFSetRef) -> CFIndex;
+    pub fn CFSetGetCountOfValue(theSet: CFSetRef, value: *const c_void) -> CFIndex;
+    pub fn CFSetGetValue(theSet: CFSetRef, value: *const c_void) -> *const c_void;
+    pub fn CFSetGetValueIfPresent(theSet: CFSetRef, candidate: *const c_void, value: *const *mut c_void) -> Boolean;
+    pub fn CFSetGetValues(theSet: CFSetRef, values: *const *mut c_void);
```
Comment (dcuddeback) [inline src/set.rs]: Here too, `values` is an out-parameter so should be `*mut *const c_void`.

File: src/set.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    pub fn CFSetApplyFunction(theSet: CFSetRef, applier: CFSetApplierFunction, context: *mut c_void);
+
+    pub fn CFSetGetTypeID() -> CFTypeID;
+
+    pub fn CFSetAddValue(theSet: CFMutableSetRef, value: *const *mut c_void);
+    pub fn CFSetRemoveAllValues(theSet: CFMutableSetRef);
+    pub fn CFSetRemoveValue(theSet: CFMutableSetRef, value: *const *mut c_void);
```
Comment (dcuddeback) [inline src/set.rs]: `value` is just a single pointer: `*const c_void`:

``` c
void CFSetRemoveValue(CFMutableSetRef theSet, const void *value);
```

**Review by dcuddeback — CHANGES_REQUESTED** [review]
Thanks @burtonageo. I left a few line comments for things that don't match the library's header files.

**Comment by burtonageo** [comment]
All done, thanks. Sorry - I sometimes get my C pointer syntax confused 😄

**Review by dcuddeback — APPROVED** [review]

**Pull request merged**